### PR TITLE
ETL Geometry V8

### PR DESCRIPTION
**Pull request by 24LopezR**
#### PR description:

This PR implements new version of ETL geometry (v8) that includes modules with several LGAD sensors (2 in this case). This geometry is integrated in MTD I17 and CMS D105.
The details about the motivation and implementation can be found in these presentations:
- [MTD DPG (10/11)](https://indico.cern.ch/event/1344170/#38-etl-geometry-update)
- [MTD DPG (13/10)](https://indico.cern.ch/event/1334483/#35-status-of-etl-geometry-and)
- [MTD DPG (09/06)](https://indico.cern.ch/event/1293576/#29-etl-geometry-update-of-sens)

#### PR validation:

For validation, a TTbar 14TeV has been run with the nominal code in D98 geometry, and the new code in D98 and D105 geometries. We observe (see attached plot (plot legend says D104 but ultimately D104->D105)):
1. Older geometries remain untouched.
2. The size of the ETL sensor is halved as expected.
3. The number of events recorded is about 10% higher, as the logical area of the pixels has been increased.

image

Added runTheMatrix test workflow 27634.0 for this scenario 2026D105.

@parbol @fabiocos

### CODE REVIEW

**Comment by fabiocos** [comment]
please test

**Comment by fabiocos** [comment]
@srimanob @bsunanda this test will not probe the new geometry, just confirm no impact on the older one. In order to explicitly test the new geometry we should add the new (D104) scenario to the short matrix. Do you agree on this addition?

**Comment by srimanob** [comment]
Hi @fabiocos 
It depends on how we will move next. Currently, we are at D98 and moving to D99. If we woudl like to jump to D101 first, or we jump to MTD first. However, both option include C18.

D99 = T32+C18+M10+I16+O9+F8
D100 = T34+C17+M11+I16+O9+F8
D101 = T34+C18+M11+I16+O9+F8
D102 = T35+C17+M11+I16+O9+F8
D103 = T34+C21+M11+I16+O9+F8
D104 = T34+C21+M11+I17+O9+F8

**Comment by fabiocos** [comment]
@24LopezR the D104 geometry configuration files need to be ri-generated by rerunning the script 

```
python3 Configuration/Geometry/scripts/generate2026Geometry.py -D104
```

Changes are unrelated to MTD, I assume updates recently introduced

**Comment by fabiocos** [comment]
@24LopezR looking at the D104 definition here, I wonder whether this si what we want for tracker and calorimeters. @srimanob @bsunanda probably we should base this on top of D99 that is the new baseline, right? The only important thing is that tracker and calo envelope are compatible with BTL and ETL respectively, and in both T32 and T34 they are, Calo envelope around ETL isn't changed to my knowledge.

**Comment by fabiocos** [comment]
@24LopezR please update D104 basing it on D99 as starting point, i.e. just D104 = T32+C18+M10+I17+O9+F8

**Comment by 24LopezR** [comment]
@fabiocos @parbol PR updated. D104 definition updated based on D99, and fixes added for failing unit tests.

**Comment by fabiocos** [comment]
@24LopezR perfect, last point: please add also workflow 27234.0 at

https://github.com/cms-sw/cmssw/blob/master/Configuration/PyReleaseValidation/python/relval_2026.py#L38

so as we may test in this PR also the new scenario (although we have no reference to compare with)

**Review by fabiocos — COMMENTED** [review]
@24LopezR I believe that there are some spurious files:

https://github.com/cms-sw/cmssw/pull/43407/files#diff-5d0fa20a04254b26dd5ea703d9b0ec80035254b273d2d6b2ebee971efc842bbe
https://github.com/cms-sw/cmssw/pull/43407/files#diff-55b12edf41539cbcaef8142574220a8ac3da6134d7bd3f6c2d7c728a9a67f9a7
https://github.com/cms-sw/cmssw/pull/43407/files#diff-4d62323e0cf4cdd50311e6b7527d33d07f2c21495d72a82e11e869ec50fbf62e

we do not need a copy for each geometry we want to test, as soon as D104 becomes our baseline we will promote the default test to that

File: Configuration/Geometry/README.md [hunk 1]
```diff
@@ -136,4 +137,4 @@ Several detector combinations have been generated:
 * D101 = T34+C18+M11+I16+O9+F8
 * D102 = T35+C17+M11+I16+O9+F8
 * D103 = T34+C21+M11+I16+O9+F8
-
+* D104 = T34+C21+M11+I17+O9+F8
```
Comment (srimanob) [inline Configuration/Geometry/README.md:140]: I think this line needs update.

**Comment by srimanob** [comment]
Hi @fabiocos 
Apart from this PR, does MTD have plan to be away from `const_cast`? 

https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-e10051/36177/llvm-analysis/runStaticChecks.log

**Comment by fabiocos** [comment]
@srimanob this was not on my radar, but we will need to address it...

**Comment by 24LopezR** [comment]
> @24LopezR I believe that there are some spurious files:
> 
> https://github.com/cms-sw/cmssw/pull/43407/files#diff-5d0fa20a04254b26dd5ea703d9b0ec80035254b273d2d6b2ebee971efc842bbe https://github.com/cms-sw/cmssw/pull/43407/files#diff-55b12edf41539cbcaef8142574220a8ac3da6134d7bd3f6c2d7c728a9a67f9a7 https://github.com/cms-sw/cmssw/pull/43407/files#diff-4d62323e0cf4cdd50311e6b7527d33d07f2c21495d72a82e11e869ec50fbf62e
> 
> we do not need a copy for each geometry we want to test, as soon as D104 becomes our baseline we will promote the default test to that

Sorry for this, I added the test files by mistake when making the commit. I'll remove them now

File: Configuration/Geometry/README.md (hunk 1 above)
Comment (24LopezR) [reply Configuration/Geometry/README.md:140]: I update this as well :)

**Comment by srimanob** [comment]
Hi @24LopezR @fabiocos 
Would you mind to wait a bit until https://github.com/cms-sw/cmssw/pull/43466 is merged, then you modify this PR to be D105. Thanks.

**Comment by fabiocos** [comment]
@24LopezR I see that the other PR was already merged, and indeed yours is no more mergeable. Please update it accordingly.

**Comment by 24LopezR** [comment]
@fabiocos I just moved D104 to D105 and generated the configuration files. I respected what's updated in PR #43466 , but the conflicts are still there. However, I believe they can easily be resolved.
Also, should I add a workflow to test D105? How should I do this?

**Comment by srimanob** [comment]
Hi @24LopezR 

It seems that you branch still have still have  
`Configuration/Geometry/python/Geometry*D104*`

**Comment by 24LopezR** [comment]
Hi @srimanob 
You are right. They are updated, but since they are not related to this PR, it is better to simply remove them here.

**Comment by srimanob** [comment]
Hi @24LopezR 
If this rebase created confusion. You can close and open a new PR on the new base. So that, everything starts from a clean plate.

**Review by fabiocos — COMMENTED** [review]
[Geometry/MTDCommonData/test/cmsDD4hepGeom.root](https://github.com/cms-sw/cmssw/pull/43407/files#diff-c3a15a35450ac96d1b08490c373cefaeef29359ab20062b0ed351b095001a2c4)

This is another spurious file that should be expunged from the release. In general do tests outside $CCMSSW_BASE/src , to avoid these issues

**Comment by fabiocos** [comment]
@24LopezR to solve the issue follow these steps (result in https://github.com/fabiocos/cmssw/tree/ruben-ETL-geom_14_0_X)

```
cmsrel CMSSW_14_0_X_2023-12-04-1100
cd CMSSW_14_0_X_2023-12-04-1100/src/
cmsenv
git cms-rebase-topic 24LopezR:ruben-ETL-geom_14_0_X
git revert 615d2fd31ef588674fee553ce9a76054290433dc
git squash 24
```
(where squash is defined in .gitconfig as
[alias]
        squash = "!f(){ git reset --soft HEAD~${1} && git commit --edit -m\"$(git log --reverse HEAD..HEAD@{1})\";};f"
)
you may keep in the commit description the whole commit history, if you like
```
git push -f my-cmssw
cd $CMSSW_BASE/..
cmsrel CMSSW_14_0_X_2023-12-06-1100  
cd CMSSW_14_0_X_2023-12-06-1100/src/
cmsenv
git cms-rebase-topic fabiocos:ruben-ETL-geom_14_0_X
```
(solve the conflicts locally)
```
git add Configuration/Geometry/README.md Configuration/Geometry/python/dict2026Geometry.py Configuration/PyReleaseValidation/python/upgradeWorkflowComponents.py Configuration/StandardSequences/python/GeometryConf.py
git rebase --continue
git push -f my-cmssw
```

**Comment by fabiocos** [comment]
There is also the possibility to use the new ```git cms-squash-topic``` command. In any case the recipe above works.

File: Configuration/PyReleaseValidation/python/upgradeWorkflowComponents.py [hunk 2]
```diff
@@ -2823,6 +2825,16 @@ def condition(self, fragment, stepList, key, hasHarvest):
         'Era' : 'Phase2C17I13M9',
         'ScenToRun' : ['GenSimHLBeamSpot','DigiTrigger','RecoGlobal', 'HARVESTGlobal', 'ALCAPhase2'],
     },
+<<<<<<< HEAD
```
Comment (srimanob) [inline Configuration/PyReleaseValidation/python/upgradeWorkflowComponents.py:2828]: I think this line should be deleted.

**Comment by srimanob** [comment]
test parameters:
- workflows = 27634.0
- relvals_opt = --what cleanedupgrade,standard,highstats,pileup,generator,extendedgen,production,identity,ged,machine,premix,nano,gpu,2017,2026

**Comment by srimanob** [comment]
@cmsbuild please test

**Review by srimanob — COMMENTED** [review]
In upgradeWorkflowComponent, Extended...104 should be 105 under D105 geometry
(Sorry for very draft comment. I confuse a bit on git on my phone)

**Comment by srimanob** [comment]
@cmsbuild please test

**Comment by bsunanda** [comment]
+geometry

**Comment by 24LopezR** [comment]
@srimanob Done! :)

**Comment by srimanob** [comment]
+Upgrade

The new workflow of D105 runs fine.

**Comment by civanch** [comment]
+1

**Comment by bsunanda** [comment]
@sunilUIET Please approve this PR

**Comment by sunilUIET** [comment]
+pdmv

**Comment by mandrenguyen** [comment]
+reconstruction
Nothing changes to comparisons, but it seems from this comment https://github.com/cms-sw/cmssw/pull/43407#issuecomment-1846791332
that the output is sensible

**Comment by fabiocos** [comment]
@cms-sw/orp-l2 the additions to the operations area are the usual set of workflows for scenario D105, and the availability of one of them in the short matrix. Do you see issues with that? Otherwise can we move forward with the integration of this PR? I have ready its extension to BTL exploiting the same geometry scenario.

**Comment by antoniovilela** [comment]
+1

**Pull request merged**